src/proto/uprotocol/uuri.rs

```diff
@@ -11,38 +11,617 @@
+            output.push_str(UUri::build_authority_part_of_uri(authority).as_str());
+        }
+        output.push('/');
+        if let Some(entity) = uri.entity.as_ref() {
+            output.push_str(UUri::build_entity_part_of_uri(entity).as_str());
+        }
+        output.push_str(UUri::build_resource_part_of_uri(uri).as_str());
+
+        // remove trailing slashes
+        Ok(Regex::new(r"/+$")
+            .unwrap()
+            .replace_all(&output, "")
+            .into_owned())
+    }
+}
+
+impl FromStr for UUri {
+    type Err = SerializationError;
+
+    /// Attempts to serialize a `String` into a `UUri`.
+    ///
+    /// # Arguments
+    ///
+    /// * `uri` - The `String` to be converted into a `UUri`.
+    ///
+    /// # Returns
+    ///
+    /// A `Result` containing either the `UUri` representation of the URI or a `SerializationError`.
+    ///
+    /// # Examples
+    ///
+    /// ```rust
+    /// use std::str::FromStr;
+    /// use up_rust::uprotocol::{UAuthority, UEntity, UResource};
+    /// use up_rust::uprotocol::UUri;
+    ///
+    /// let uri = UUri {
+    ///     entity: Some(UEntity {
+    ///         name: "example.com".to_string(),
+    ///         ..Default::default()
+    ///     })
+    ///     .into(),
+    ///     resource: Some(UResource {
+    ///         name: "rpc".to_string(),
+    ///         instance: Some("raise".to_string()),
+    ///         ..Default::default()
+    ///     })
+    ///     .into(),
+    ///     authority: None.into(),
+    ///     ..Default::default()
+    /// };
+    ///
+    /// let uri_from = UUri::from_str("/example.com//rpc.raise").unwrap();
+    /// assert_eq!(uri, uri_from);
+    /// ````
+    fn from_str(uri: &str) -> Result<Self, Self::Err> {
+        if uri.is_empty() {
+            return Err(SerializationError::new("URI is empty"));
+        }
+
+        let uri = if let Some(index) = uri.find(':') {
+            uri[index + 1..].to_string()
+        } else {
+            uri.replace('\\', "/")
+        };
+        let is_local: bool = !uri.starts_with("//");
+        let uri_parts = Self::pattern_split(&uri, "/");
+
+        if uri_parts.len() < 2 {
+            return Err(SerializationError::new("URI is invalid"));
+        }
+
+        #[allow(unused_assignments)]
+        let mut name: String = String::default();
+        let mut version: String = String::default();
+        let mut resource: Option<UResource> = None;
+        let mut authority: Option<UAuthority> = None;
+
+        if is_local {
+            name = uri_parts[1].to_string();
+            if uri_parts.len() > 2 {
+                version = uri_parts[2].to_string();
+            }
+            if uri_parts.len() > 3 {
+                resource = Some(UResource::from(uri_parts[3].as_str()));
+            }
+        } else {
+            if uri_parts.len() > 2 {
+                if uri_parts[2].trim().is_empty() {
+                    return Err(SerializationError::new("URI is invalid"));
```

Comment:
I think here the error message could be made a bit more specific, like this:
```rust
return Err(SerializationError::new("Remote URI missing UAuthority"));
```
WDYT?